src/resolver.ts

```diff
@@ -181,16 +185,28 @@ export class Resolver extends DiagnosticEmitter {
     ctxElement: Element,
     /** Contextual types, i.e. `T`. */
     ctxTypes: Map<string,Type> | null = null,
+    /** Contextual type definitions */
+    ctxTypeDefinitions: Map<string,TypeDefinition> | null = null,
```

Comment:
I'm looking at these two fields and it looks like they could be combined into one, for example:
```ts
    ctxTypesAndDefinitions: Map<string,TypeAndTypeDefinition> | null = null,
```
where `TypeAndTypeDefinition`:
```ts
class TypeAndTypeDefinition {
  constructor(
    public type: Type,
    public typeDefinition: TypeDefinition | null = null
  ) {}
}
```
Also `TypeDefinition` already aggregate `type`. So in theory, we could just always use `TypeDefinition` instead `TypeAndTypeDefinition`. @dcodeIO wdyt?

Reply:
I feel like all of my code here is a big hack, but replacing `ctxTypes` with `ctxTypeDefinitions` seems fine, because the original code (now in `resolveTypeDefinition`) already checked if the `TypeDefinition` had a resolved type (for builtins, according to a comment).

If `ctxTypes` was replaced with `ctxTypeDefinitions`, and the corresponding `TypeDefinitions` were resolved, then that check should catch those types. You can even compare lines 211-229 with 324-342; they are exactly the same in functionality.